### PR TITLE
Integrate Product features form

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Integrate Product features form
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | ~
| How to test?      | CI tests green
| Possible impacts? | ~

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23159)

### CODE REVIEW

File: tests/Unit/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProviderTest.php [hunk 1]
```diff
@@ -618,6 +697,8 @@ private function getDefaultOutputData(): array
                 'reference' => 'reference',
             ],
             'suppliers' => [],
+            'features' => [
+            ],
```
Comment (Progi1984) [inline tests/Unit/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProviderTest.php]:
```suggestion
            'features' => [],
```

File: tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/FeatureValuesCommandBuilderTest.php [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+class FeatureValuesCommandBuilderTest extends AbstractProductCommandBuilderTest
+{
+    /**
+     * @dataProvider getExpectedCommands
+     *
+     * @param array $formData
+     * @param array $expectedCommands
+     */
+    public function testBuildCommand(array $formData, array $expectedCommands)
```
Comment (Progi1984) [inline tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/FeatureValuesCommandBuilderTest.php]:
```suggestion
    public function testBuildCommand(array $formData, array $expectedCommands): void
```

File: tests/Integration/PrestaShopBundle/Form/FormClonerTest.php [hunk 3]
```diff
@@ -145,6 +196,57 @@ public function testCloneWithOptions()
         ]);
     }
 
+    /**
+     * This is based on a use case of duplicating a choice type, which internally builds its own
+     * view transformers based on the options. So we need to check that cloning the form didn't duplicate
+     * this transformer but instead replaced it with the appropriate options.
+     */
+    public function testOverrideViewTransformers()
```
Comment (Progi1984) [inline tests/Integration/PrestaShopBundle/Form/FormClonerTest.php]:
```suggestion
    public function testOverrideViewTransformers(): void
```

File: tests/Integration/PrestaShopBundle/Form/FormClonerTest.php [hunk 4]
```diff
@@ -97,6 +127,27 @@ public function testCloneWithViewTransformer()
         $this->assertContains($transformer, $clonedTargetFormTransformers);
     }
 
+    public function testCloneChildWithViewTransformer()
```
Comment (Progi1984) [inline tests/Integration/PrestaShopBundle/Form/FormClonerTest.php]:
```suggestion
    public function testCloneChildWithViewTransformer(): void
```

File: tests/Integration/PrestaShopBundle/Form/FormClonerTest.php [hunk 5]
```diff
@@ -78,6 +84,27 @@ public function testCloneWithModelTransformer()
         $this->assertContains($transformer, $clonedTargetFormTransformers);
     }
 
+    public function testCloneChildWithModelTransformer()
```
Comment (Progi1984) [inline tests/Integration/PrestaShopBundle/Form/FormClonerTest.php]:
```suggestion
    public function testCloneChildWithModelTransformer(): void
```

File: src/Adapter/Feature/Repository/FeatureValueRepository.php [hunk 6]
```diff
@@ -206,21 +228,34 @@ private function getFeatureValueLocalizedValues(int $featureValueId): array
...
+        foreach ($filters as $key => $value) {
+            switch ($key) {
+                case 'id_product':
+                    $qb
+                        ->andWhere('fp.id_product = :productId')
+                        ->setParameter('productId', (int) $value)
+                    ;
+                break;
+                default:
+                    $qb
```
Comment (matks) [inline src/Adapter/Feature/Repository/FeatureValueRepository.php:262]: Should we use a secure list of authorized filters like we did for example in https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Query/CmsPageCategoryQueryBuilder.php#L115 ?

File: src/Adapter/Form/ChoiceProvider/FeatureValuesChoiceProvider.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    public function getChoices(array $options)
+    {
+        if (empty($options['feature_id'])) {
+            return [];
+        }
+
+        $featureId = (int) $options['feature_id'];
+        $filters = [
+            'id_feature' => $featureId,
```
Comment (matks) [inline src/Adapter/Form/ChoiceProvider/FeatureValuesChoiceProvider.php]: I think @PierreRambaud will tell you 😄  :
```suggestion
            'id_feature' => (int) $options['feature_id'],
```

File: src/Adapter/Form/ChoiceProvider/FeatureValuesChoiceProvider.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            $filters['custom'] = $options['custom'];
+        }
+
+        $featureValues = $this->featureValueRepository->getFeatureValues(null, null, $filters);
+        $choices = [];
+        foreach ($featureValues as $feature) {
+            if (!empty($feature['localized_values'][$this->contextLanguageId])) {
+                $featureValueName = $feature['localized_values'][$this->contextLanguageId];
+            } else {
+                $featureValueName = $feature['localized_values'][$this->defaultLanguageId];
```
Comment (matks) [inline src/Adapter/Form/ChoiceProvider/FeatureValuesChoiceProvider.php:84]: Should we handle the case where `$feature['localized_values'][$this->defaultLanguageId]` is null? (so handle the case where SQL data comes as invalid or incomplete)?

File: src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/FeatureValueListener.php [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+use PrestaShopBundle\Form\FormCloner;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+
+/**
+ * This listener dynamically updates the choices allowed in the feature value selector,
+ * choices are fetched based on the selected feature.
+ */
+class FeatureValueListener implements EventSubscriberInterface
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/FeatureValueListener.php:41]: Could you add an unit test for the listener, please ?

File: src/Adapter/Product/FeatureValue/Update/ProductFeatureValueUpdater.php [hunk 10]
```diff
@@ -113,10 +115,18 @@ public function setFeatureValues(ProductId $productId, array $productFeatureValu
...
         foreach ($productFeatureValues as $productFeatureValue) {
             $this->featureRepository->assertExists($productFeatureValue->getFeatureId());
             if (null !== $productFeatureValue->getFeatureValueId()) {
-                $this->featureValueRepository->assertExists($productFeatureValue->getFeatureValueId());
+                $featureValue = $this->featureValueRepository->get($productFeatureValue->getFeatureValueId());
+                if ((int) $featureValue->id_feature !== $productFeatureValue->getFeatureId()->getValue()) {
+                    throw new InvalidAssociatedFeatureException('You cannot associate a value to another feature.');
+                }
+                if (in_array($productFeatureValue->getFeatureValueId()->getValue(), $previousFeatureIds)) {
+                    throw new DuplicateFeatureValueAssociationException('You cannot associate the same feature value more than once.');
```
Comment (matks) [inline src/Adapter/Product/FeatureValue/Update/ProductFeatureValueUpdater.php:127]: 👍

File: src/Core/Form/IdentifiableObject/CommandBuilder/Product/FeatureValuesCommandBuilder.php [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+    /**
+     * @param array $featureValuesData
+     *
+     * @return array
+     */
+    private function formatFeatureValues(array $featureValuesData): array
+    {
+        $featureValues = [];
+        foreach ($featureValuesData as $featureValueDatum) {
```
Comment (matks) [inline src/Core/Form/IdentifiableObject/CommandBuilder/Product/FeatureValuesCommandBuilder.php:67]: Dada, dadum 🎶

**Review by Progi1984 — COMMENTED** [review]
Small feedbacks but good job

**Review by matks — COMMENTED** [review]
Reviewed half of the PR, will do the other half later 😄

File: src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/FeatureValueListener.php [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * @param FormEvent $event
+     */
+    public function updateFeatureValuesOptions(FormEvent $event): void
+    {
+        $form = $event->getForm();
+        $data = $event->getData();
+
+        if (empty($data) || empty($data['feature_id'])) {
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/FeatureValueListener.php]:
```suggestion
        if (empty($data['feature_id'])) {
```

File: src/Adapter/Feature/Repository/FeatureValueRepository.php (hunk 6 above)
Comment (jolelievre) [reply src/Adapter/Feature/Repository/FeatureValueRepository.php:262]: It was a very basic rough implementation I didn't put much thought into it 😅  But yes, your suggestion will definitely improve it

File: src/Adapter/Form/ChoiceProvider/FeatureValuesChoiceProvider.php (hunk 8 above)
Comment (jolelievre) [reply src/Adapter/Form/ChoiceProvider/FeatureValuesChoiceProvider.php:84]: I guess I can check if it exists, if not return first value no matter which language it matches it's still better than empty string

File: src/Core/Form/IdentifiableObject/CommandBuilder/Product/FeatureValuesCommandBuilder.php (hunk 11 above)
Comment (jolelievre) [reply src/Core/Form/IdentifiableObject/CommandBuilder/Product/FeatureValuesCommandBuilder.php:67]: Yeah sometimes I let PHPStorm decide with such variables name to please it 😅

File: src/PrestaShopBundle/Form/Admin/Sell/Product/FeaturesType.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        $builder
+            ->add('feature_values', CollectionType::class, [
+                'entry_type' => FeatureValueType::class,
+                'required' => false,
+                'allow_add' => true,
+                'allow_delete' => true,
+                'prototype_name' => '__FEATURE_VALUE_INDEX__',
+            ])
+            ->add('add_feature', IconButtonType::class, [
+                'label' => $this->trans('Add a feature', 'Admin.Catalog.Feature'),
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/FeaturesType.php:52]: 'Add a new feature' also exists, wouldn't it be better?

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php [hunk 14]
```diff
@@ -182,6 +184,14 @@ private function getErrorMessages(Exception $e): array
                     'Admin.Notifications.Error'
                 ),
             ],
+            DuplicateFeatureValueAssociationException::class => $this->trans(
+                'You cannot associate the same feature value more than once.',
+                'Admin.Notifications.Error'
+            ),
+            InvalidAssociatedFeatureException::class => $this->trans(
+                'You cannot associate a feature value to another feature.',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: I'm not sure to understand what you mean here?

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: I'm not fully satisfied either 😅 It's not supposed to happen but the idea is that you cannot associate a feature value that is associated to Feature A to another Feature B. For example in default data we have two features:

- Composition
  - Polyester
  - Wool
  - Ceramic
  - ...
- Property
  - Long sleeves
  - Short sleeves
  - ...

So you can't associate the value "Ceramic" to the feature "Property" It is not supposed to happen anyway since when you select a feature, the possible values are automatically updated with appropriate values. I guess it could happen if you have edited the value while you were selecting a feature, and you save after it's been modified.

Maybe we could rephrase it to something like `The property you selected does not belong to the selected feature.` ?

File: src/PrestaShopBundle/Form/Admin/Sell/Product/FeaturesType.php (hunk 13 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/FeaturesType.php:52]: I think `Add a new feature` is used in the Feature list where you can create the pre-defined features. Here it's more `Add a feature to the product` It's an association more than a creation, but I used the same wording as in the previous page though

Comment (Julievrz) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/FeaturesType.php:52]: Oops, yes sorry I wasn't looking at the good wording. Thank you ;)

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php (hunk 14 above)
Comment (Julievrz) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:192]: Thank you it is way clearer now.
What do you think of: 'The selected value already belongs to another feature.'?

Comment (Julievrz) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: Thank you, it is clearer now! :)
What do you think of: 
'The selected value already belongs to another feature.'
or
'A value cannot be associated with two different features.'

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: MM, I think the first one is more suited but I would drop the already `The selected value belongs to another feature.` because it sort of implies the user was moving the value to another feature, or it didn't have a feature yet Whereas it is trying to assign it to a product 🤔

Comment (Julievrz) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: Sounds good to me!

File: src/Adapter/Form/ChoiceProvider/FeatureValuesChoiceProvider.php [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+namespace PrestaShop\PrestaShop\Adapter\Form\ChoiceProvider;
+
+use PrestaShop\PrestaShop\Adapter\Feature\Repository\FeatureValueRepository;
+use PrestaShop\PrestaShop\Adapter\LegacyContext;
+use PrestaShop\PrestaShop\Core\Form\ConfigurableFormChoiceProviderInterface;
+
+class FeatureValuesChoiceProvider implements ConfigurableFormChoiceProviderInterface
+{
+    private $featureValueRepository;
```
Comment (Progi1984) [inline src/Adapter/Form/ChoiceProvider/FeatureValuesChoiceProvider.php:40]:
```suggestion
    /**
     * @var FeatureValueRepository
     */
    private $featureValueRepository;
```

**Review by Progi1984 — COMMENTED** [review]
A small feedback

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php (hunk 14 above)
Comment (Julievrz) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]:
```suggestion
                'The selected value belongs to another feature.',
```

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by matks — APPROVED** [review]
